File: client/connection_log.py
"""Local-only connection log for product clients (device storage, user-exportable).

Events stay on the device under the product data directory. This module never
uploads log content to the node, VPN APP Shop, or any remote collector.

Support handoff: the user exports a plain-text file (Settings → Export log) and
emails that file to support themselves. Diagnostics in the export are for that
manual handoff only — not auto-telemetry.
"""

from __future__ import annotations

import json
import os
import sys
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Iterable, Mapping, Optional

LOG_FILENAME = "connection_log.jsonl"
DEFAULT_MAX_EVENTS = 500

# Kinds used by the product UI / Settings surface.
KIND_CONNECT = "connect"
KIND_DISCONNECT = "disconnect"
KIND_SESSION = "session"
KIND_ERROR = "error"
KIND_INFO = "info"
KIND_LEAK_TEST = "leak_test"
KIND_SETTINGS = "settings"

# Keys commonly useful for support (safe, no secrets).
_DIAG_PREFERRED_KEYS = (
    "product",
    "client_version",
    "platform",
    "os_name",
    "python",
    "outcome",
    "error",
    "error_code",
    "residual_host",
    "residual_port",
    "entry_country",
    "multihop",
    "session_vpn_ip",
    "residual_capture",
)


@dataclass(frozen=True)
class ConnectionLogEvent:
    """One local connection-log entry (optional support diagnostics in *detail*)."""

    ts: float
    kind: str
    message: str
    detail: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {
            "ts": float(self.ts),
            "kind": str(self.kind),
            "message": str(self.message),
        }
        if self.detail:
            # Only JSON-serializable scalars / short strings
            clean: dict[str, Any] = {}
            for k, v in self.detail.items():
                key = str(k).strip()
                if not key:
                    continue
                if isinstance(v, (bool, int, float)):
                    clean[key] = v
                elif v is None:
                    continue
                else:
                    s = str(v).strip()
                    if s:
                        clean[key] = s[:500]
            if clean:
                out["detail"] = clean
        return out

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "ConnectionLogEvent":
        raw_detail = data.get("detail")
        detail: dict[str, Any] = {}
        if isinstance(raw_detail, dict):
            detail = {str(k): v for k, v in raw_detail.items()}
        return cls(
            ts=float(data.get("ts") or 0.0),
            kind=str(data.get("kind") or KIND_INFO),
            message=str(data.get("message") or ""),
            detail=detail,
        )

    def format_line(self) -> str:
        """Human-readable single line for export / Settings list."""
        try:
            stamp = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(self.ts))
        except (OverflowError, OSError, ValueError):
            stamp = str(self.ts)
        base = f"[{stamp}] {self.kind}: {self.message}"
        if not self.detail:
            return base
        # Prefer compact key=value tails for support scanning
        parts: list[str] = []
        for key in _DIAG_PREFERRED_KEYS:
            if key in self.detail and key not in ("product", "client_version", "platform", "os_name", "python"):
                parts.append(f"{key}={self.detail[key]}")
        # Any remaining custom keys (skip base diag already in export header)
        skip = set(_DIAG_PREFERRED_KEYS[:5])
        for key, val in self.detail.items():
            if key in skip:
                continue
            if key in _DIAG_PREFERRED_KEYS and any(p.startswith(f"{key}=") for p in parts):
                continue
            if key not in _DIAG_PREFERRED_KEYS:
                parts.append(f"{key}={val}")
        if not parts:
            return base
        return base + " | " + " ".join(parts)


def product_client_version() -> str:
    """Shipped client VERSION string (no network)."""
    try:
        p = Path(__file__).resolve().parent / "VERSION"
        if p.is_file():
            v = p.read_text(encoding="utf-8").strip().splitlines()[0].strip()
            if v:
                return v
    except OSError:
        pass
    return "unknown"


def build_support_diagnostics(
    *,
    extra: Mapping[str, Any] | None = None,
    include_runtime: bool = True,
) -> dict[str, Any]:
    """Small diagnostic snapshot for support export / event detail.

    Pure local facts only — never opens network sockets or calls remote APIs.
    """
    snap: dict[str, Any] = {
        "product": "Restore Privacy",
        "client_version": product_client_version(),
    }
    if include_runtime:
        snap["platform"] = str(sys.platform)
        snap["os_name"] = str(os.name)
        snap["python"] = (
            f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}"
        )
    if extra:
        for k, v in extra.items():
            key = str(k).strip()
            if not key or v is None:
                continue
            if isinstance(v, (bool, int, float)):
                snap[key] = v
            else:
                s = str(v).strip()
                if s:
                    snap[key] = s[:500]
    return snap


def log_dir() -> Path:
    """Product local data directory (same family as Windows settings_store)."""
    if os.name == "nt":
        base = os.environ.get("LOCALAPPDATA") or str(
            Path.home() / "AppData" / "Local"
        )
        return Path(base) / "RestorePrivacy"
    xdg = os.environ.get("XDG_DATA_HOME")
    if xdg:
        return Path(xdg) / "restore-privacy"
    return Path.home() / ".local" / "share" / "restore-privacy"


def default_log_path() -> Path:
    return log_dir() / LOG_FILENAME


def append_event(
    kind: str,
    message: str,
    *,
    path: Optional[Path] = None,
    ts: Optional[float] = None,
    max_events: int = DEFAULT_MAX_EVENTS,
    detail: Mapping[str, Any] | None = None,
    include_diagnostics: bool = True,
) -> ConnectionLogEvent:
    """Append one event to the local JSONL log; trim oldest if over max_events.

    When *include_diagnostics* is true (default), merges a support diagnostic
    snapshot (version / platform) with any caller *detail* (endpoint, error, …).
    """
    merged: dict[str, Any] = {}
    if include_diagnostics:
        merged.update(build_support_diagnostics())
    if detail:
        for k, v in detail.items():
            key = str(k).strip()
            if not key or v is None:
                continue
            if isinstance(v, (bool, int, float)):
                merged[key] = v
            else:
                s = str(v).strip()
                if s:
                    merged[key] = s[:500]
    event = ConnectionLogEvent(
        ts=float(ts if ts is not None else time.time()),
        kind=str(kind or KIND_INFO),
        message=str(message or "").strip() or "(empty)",
        detail=merged,
    )
    p = path or default_log_path()
    p.parent.mkdir(parents=True, exist_ok=True)
    with p.open("a", encoding="utf-8") as fh:
        fh.write(json.dumps(event.to_dict(), ensure_ascii=False) + "\n")
    if max_events > 0:
        _trim_log(p, max_events=max_events)
    return event


def read_events(
    *,
    path: Optional[Path] = None,
    limit: Optional[int] = None,
) -> list[ConnectionLogEvent]:
    """Read events from the local log (oldest first). ``limit`` keeps the newest N."""
    p = path or default_log_path()
    if not p.is_file():
        return []
    events: list[ConnectionLogEvent] = []
    try:
        text = p.read_text(encoding="utf-8")
    except OSError:
        return []
    for line in text.splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            data = json.loads(line)
            if isinstance(data, dict):
                events.append(ConnectionLogEvent.from_dict(data))
        except (json.JSONDecodeError, TypeError, ValueError):
            continue
    if limit is not None and limit >= 0 and len(events) > limit:
        events = events[-limit:]
    return events


def format_export(
    events: Optional[Iterable[ConnectionLogEvent]] = None,
    *,
    path: Optional[Path] = None,
    diagnostics: Mapping[str, Any] | None = None,
) -> str:
    """Plain-text export body (user can save/email). Never leaves the device here."""
    if events is None:
        events = read_events(path=path)
    event_list = list(events)
    snap = build_support_diagnostics(extra=diagnostics)
    header_lines = [
        "# Restore Privacy connection log (local only)",
        "# Not uploaded by the client. User-exported file for support handoff.",
        "# Support: email this file to support yourself (no automatic upload).",
        f"# product={snap.get('product')} client_version={snap.get('client_version')}",
        f"# platform={snap.get('platform')} os_name={snap.get('os_name')} "
        f"python={snap.get('python')}",
        "# --- events ---",
    ]
    lines = [ev.format_line() for ev in event_list]
    body = "\n".join(header_lines) + "\n"
    if lines:
        body += "\n".join(lines) + "\n"
    return body


def export_to_file(
    dest: Path,
    *,
    source: Optional[Path] = None,
    events: Optional[Iterable[ConnectionLogEvent]] = None,
    diagnostics: Mapping[str, Any] | None = None,
) -> Path:
    """Write export text to ``dest``; returns path written."""
    body = format_export(events, path=source, diagnostics=diagnostics)
    dest = Path(dest)
    dest.parent.mkdir(parents=True, exist_ok=True)
    dest.write_text(body, encoding="utf-8")
    return dest


def clear_events(*, path: Optional[Path] = None) -> None:
    """Erase the local log file (user-initiated clear)."""
    p = path or default_log_path()
    try:
        if p.is_file():
            p.unlink()
    except OSError:
        pass


def log_module_has_no_network_upload() -> bool:
    """Structural honesty: this module must not import network clients.

    Used by unit tests. Returns True when the source has no upload/HTTP client
    imports. Does not scan the whole process — only this file's source.
    """
    import ast

    src = Path(__file__).read_text(encoding="utf-8")
    try:
        tree = ast.parse(src)
    except SyntaxError:
        return False
    banned_roots = {
        "urllib",
        "requests",
        "httpx",
        "socket",
        "http",
        "aiohttp",
        "ftplib",
    }
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            for alias in node.names:
                root = alias.name.split(".", 1)[0]
                if root in banned_roots:
                    return False
        elif isinstance(node, ast.ImportFrom):
            if node.module:
                root = node.module.split(".", 1)[0]
                if root in banned_roots:
                    return False
    # Docstring / export header must state local-only / not uploaded
    low = src.lower()
    return (
        "never" in low
        and "upload" in low
        and ("not uploaded" in low or "local only" in low or "local-only" in low)
    )


def _trim_log(path: Path, *, max_events: int) -> None:
    events = read_events(path=path)
    if len(events) <= max_events:
        return
    keep = events[-max_events:]
    lines = [json.dumps(ev.to_dict(), ensure_ascii=False) for ev in keep]
    path.write_text("\n".join(lines) + "\n", encoding="utf-8")

File: client_app/lib/connection_log.dart
/// Local-only connection log (device storage, user-exportable).
///
/// Events are not uploaded by the client. Export produces plain text the user
/// can save or email to support themselves. Diagnostics (version / platform /
/// outcome) are for that manual handoff only.
library;

import 'dart:io' show Platform;

const String kLogKindConnect = 'connect';
const String kLogKindDisconnect = 'disconnect';
const String kLogKindSession = 'session';
const String kLogKindError = 'error';
const String kLogKindInfo = 'info';
const String kLogKindLeakTest = 'leak_test';

const int kDefaultMaxEvents = 500;

class ConnectionLogEvent {
  final double ts;
  final String kind;
  final String message;
  final Map<String, String> detail;

  const ConnectionLogEvent({
    required this.ts,
    required this.kind,
    required this.message,
    this.detail = const {},
  });

  Map<String, dynamic> toJson() {
    final m = <String, dynamic>{
      'ts': ts,
      'kind': kind,
      'message': message,
    };
    if (detail.isNotEmpty) {
      m['detail'] = detail;
    }
    return m;
  }

  factory ConnectionLogEvent.fromJson(Map<String, dynamic> data) {
    final raw = data['detail'];
    final d = <String, String>{};
    if (raw is Map) {
      raw.forEach((k, v) {
        if (v != null) d[k.toString()] = v.toString();
      });
    }
    return ConnectionLogEvent(
      ts: (data['ts'] is num) ? (data['ts'] as num).toDouble() : 0.0,
      kind: data['kind']?.toString() ?? kLogKindInfo,
      message: data['message']?.toString() ?? '',
      detail: d,
    );
  }

  String formatLine() {
    final dt = DateTime.fromMillisecondsSinceEpoch((ts * 1000).round(), isUtc: false);
    final stamp =
        '${dt.year.toString().padLeft(4, '0')}-${dt.month.toString().padLeft(2, '0')}-${dt.day.toString().padLeft(2, '0')} '
        '${dt.hour.toString().padLeft(2, '0')}:${dt.minute.toString().padLeft(2, '0')}:${dt.second.toString().padLeft(2, '0')}';
    var base = '[$stamp] $kind: $message';
    if (detail.isEmpty) return base;
    final parts = <String>[];
    for (final key in ['outcome', 'error', 'error_code', 'residual_host', 'session_vpn_ip', 'residual_capture']) {
      if (detail.containsKey(key)) parts.add('$key=${detail[key]}');
    }
    detail.forEach((k, v) {
      if (['product', 'client_version', 'platform', 'os_name'].contains(k)) return;
      if (parts.any((p) => p.startsWith('$k='))) return;
      parts.add('$k=$v');
    });
    if (parts.isEmpty) return base;
    return '$base | ${parts.join(' ')}';
  }
}

/// Injectable store so unit tests drive the real [ConnectionLog] path.
abstract class ConnectionLogBackend {
  Future<List<String>> readLines();
  Future<void> writeLines(List<String> lines);
}

/// In-memory backend (tests + ephemeral sessions).
class MemoryConnectionLogBackend implements ConnectionLogBackend {
  MemoryConnectionLogBackend([List<String>? seed]) : _lines = List.of(seed ?? const []);

  final List<String> _lines;

  @override
  Future<List<String>> readLines() async => List.of(_lines);

  @override
  Future<void> writeLines(List<String> lines) async {
    _lines
      ..clear()
      ..addAll(lines);
  }
}

/// Support diagnostic snapshot (local only — no network).
Map<String, String> buildSupportDiagnostics({
  String clientVersion = 'unknown',
  String platform = 'flutter',
  Map<String, String>? extra,
}) {
  final snap = <String, String>{
    'product': 'Restore Privacy',
    'client_version': clientVersion.isEmpty ? 'unknown' : clientVersion,
    'platform': platform.isEmpty ? 'flutter' : platform,
  };
  if (extra != null) {
    snap.addAll(extra);
  }
  return snap;
}

/// Failure line for local connection-log export (support handoff).
///
/// Prefer residual-honest UI / native status over a bare `Connect failed` token
/// so exports match what the user saw and are actionable for support.
String connectionLogConnectFailureMessage(
  String? uiOrNativeStatus, {
  String fallback = 'Connect failed',
}) {
  final s = (uiOrNativeStatus ?? '').trim();
  if (s.isEmpty) return fallback;
  final low = s.toLowerCase();
  // Connecting progress is not a terminal failure detail by itself
  if (low.startsWith('connecting') || low.contains('still connecting')) {
    return '$fallback — tunnel did not complete (last status: $s)';
  }
  // Already a bare failure with no extra detail
  if (low == 'connect failed' || low == fallback.toLowerCase()) {
    return s;
  }
  // Prefer residual-honest status as the error body (do not prefix if already detailed)
  if (low.contains('connect failed')) return s;
  return s;
}

/// Platform label for connection-log / support export (macOS vs iOS vs other).
String connectionLogPlatformLabel() {
  if (Platform.isMacOS) return 'macos';
  if (Platform.isIOS) return 'ios';
  if (Platform.isAndroid) return 'android';
  return 'flutter';
}

class ConnectionLog {
  ConnectionLog(
    this.backend, {
    this.maxEvents = kDefaultMaxEvents,
    this.clientVersion = 'unknown',
    this.platformLabel = 'flutter',
  });

  final ConnectionLogBackend backend;
  final int maxEvents;
  final String clientVersion;
  final String platformLabel;

  Future<ConnectionLogEvent> appendEvent(
    String kind,
    String message, {
    double? ts,
    Map<String, String>? detail,
    bool includeDiagnostics = true,
  }) async {
    final merged = <String, String>{};
    if (includeDiagnostics) {
      merged.addAll(buildSupportDiagnostics(
        clientVersion: clientVersion,
        platform: platformLabel,
      ));
    }
    if (detail != null) merged.addAll(detail);
    final event = ConnectionLogEvent(
      ts: ts ?? DateTime.now().millisecondsSinceEpoch / 1000.0,
      kind: kind.isEmpty ? kLogKindInfo : kind,
      message: message.trim().isEmpty ? '(empty)' : message.trim(),
      detail: merged,
    );
    final lines = await backend.readLines();
    lines.add(_encodeLine(event));
    final trimmed = maxEvents > 0 && lines.length > maxEvents
        ? lines.sublist(lines.length - maxEvents)
        : lines;
    await backend.writeLines(trimmed);
    return event;
  }

  Future<List<ConnectionLogEvent>> readEvents({int? limit}) async {
    final lines = await backend.readLines();
    final events = <ConnectionLogEvent>[];
    for (final line in lines) {
      final e = _decodeLine(line);
      if (e != null) events.add(e);
    }
    if (limit != null && limit >= 0 && events.length > limit) {
      return events.sublist(events.length - limit);
    }
    return events;
  }

  /// Plain-text export body (local only — not uploaded by the client).
  Future<String> formatExport({int? limit}) async {
    final events = await readEvents(limit: limit);
    final snap = buildSupportDiagnostics(
      clientVersion: clientVersion,
      platform: platformLabel,
    );
    final buf = StringBuffer()
      ..writeln('# Restore Privacy connection log (local only)')
      ..writeln('# Not uploaded by the client. User-exported file for support handoff.')
      ..writeln('# Support: email this file to support yourself (no automatic upload).')
      ..writeln('# product=${snap['product']} client_version=${snap['client_version']}')
      ..writeln('# platform=${snap['platform']}')
      ..writeln('# --- events ---');
    for (final e in events) {
      buf.writeln(e.formatLine());
    }
    return buf.toString();
  }

  Future<void> clear() async => backend.writeLines([]);

  String _encodeLine(ConnectionLogEvent e) {
    // Minimal JSON without extra deps.
    final msg = e.message
        .replaceAll('\\', '\\\\')
        .replaceAll('"', '\\"')
        .replaceAll('\n', '\\n');
    final parts = <String>[
      '"ts":${e.ts}',
      '"kind":"${e.kind}"',
      '"message":"$msg"',
    ];
    if (e.detail.isNotEmpty) {
      final dparts = e.detail.entries.map((en) {
        final k = en.key.replaceAll('\\', '\\\\').replaceAll('"', '\\"');
        final v = en.value.replaceAll('\\', '\\\\').replaceAll('"', '\\"');
        return '"$k":"$v"';
      }).join(',');
      parts.add('"detail":{$dparts}');
    }
    return '{${parts.join(',')}}';
  }

  ConnectionLogEvent? _decodeLine(String line) {
    final t = line.trim();
    if (t.isEmpty) return null;
    try {
      // Tiny parser for our fixed shape.
      final tsMatch = RegExp(r'"ts"\s*:\s*([0-9.]+)').firstMatch(t);
      final kindMatch = RegExp(r'"kind"\s*:\s*"([^"]*)"').firstMatch(t);
      final msgMatch = RegExp(r'"message"\s*:\s*"((?:\\.|[^"\\])*)"').firstMatch(t);
      if (tsMatch == null || kindMatch == null || msgMatch == null) return null;
      final msg = msgMatch.group(1)!
          .replaceAll('\\n', '\n')
          .replaceAll('\\"', '"')
          .replaceAll('\\\\', '\\');
      final detail = <String, String>{};
      final detailBlock = RegExp(r'"detail"\s*:\s*\{([^}]*)\}').firstMatch(t);
      if (detailBlock != null) {
        final body = detailBlock.group(1)!;
        for (final m in RegExp(r'"([^"]+)"\s*:\s*"((?:\\.|[^"\\])*)"').allMatches(body)) {
          detail[m.group(1)!] = m
              .group(2)!
              .replaceAll('\\n', '\n')
              .replaceAll('\\"', '"')
              .replaceAll('\\\\', '\\');
        }
      }
      return ConnectionLogEvent(
        ts: double.parse(tsMatch.group(1)!),
        kind: kindMatch.group(1)!,
        message: msg,
        detail: detail,
      );
    } catch (_) {
      return null;
    }
  }
}
